docs/architecture-decisions/flag-type.md

```diff
@@ -0,0 +1,127 @@
+
+The root cause of this issue is the absence of a dedicated, authoritative type definition for the flag itself. Instead, the type is inferred from the `defaultVariant` or API itself (`T` from `Resolve<T>`) , which is not always a reliable source of truth for all possible variants. This problem is getting worse by the planned support for code-defined defaults (as detailed in the [Support Code Default ADR](https://github.com/open-feature/flagd/blob/main/docs/architecture-decisions/support-code-default.md)), which allows the `defaultVariant` to be `null`. This makes it impossible to resolve the flag's type from the `defaultVariant`, increasing the risk of runtime errors.
+
+
+## Requirements
+
+* The new `flagType`field in the flag definition must be optional to ensure backward compatibility.
+* If the `flagType`field is present, `flagd` must validate that all variant values of the flag conform to this type during initialization.
+* Type mismatches found during initialization must be reported as errors.
+* The `Resolve<T>` and `ResolveAll` methods must use the `flagType`field for validation when it is available.
+* The implementation must be consistent with the OpenFeature specification and the flag manifest schema.
+
+
+## Considered Options
+
+* **Consistent `defaultVariant` Validation:** Align the behavior of `Resolve<T>` with `ResolveAll` by making `Resolve<T>` validate the evaluated variant against the type of the `defaultVariant`.
+* **API Extension with Explicit Flag Type:** Introduce an optional `flagType`property to the flag definition to serve as the single source for type validation.
+
+
+## Proposal
+
+This proposal is to extend the flag definition with an optional `flagType`property. This approach is chosen over simply aligning the `Resolve<T>` and `ResolveAll` validation because it addresses the root cause of the type inconsistency and provides a more robust, long-term solution.
+
+By introducing an explicit `flagType`field, it establishes a single source of truth for the flag's type, independent of its variants. This allows for early and consistent type validation during flag definition parsing, preventing type-related errors at runtime.
+
+The new `flagType`field will be optional to maintain backward compatibility with existing flag configurations. If the field is omitted, `flagd` will treat the flag as having `object`, and no type validation will be performed against the `defaultVariant`. When the `flagType`field is present, `flagd` will enforce that all variants of the flag conform to the specified type.
+
+This change will make the behavior of `flagd` more predictable and reliable.
+
+
+### API changes
+
+The `flagd` flag definition will be updated to include an optional `flagType`property. This property will be a string enum with the following possible values: `"boolean"`, `"string"`, `"integer"`, `"float"`, and `"object"`. This aligns with the OpenFeature CLI and the flag manifest schema.
```

Comment:
I'm not 100% sure it's a good idea to differentiate between float/int for this purpose. A couple reasons:

- I think we could easily write a JSON schema to enforce that flags with `"flagType":boolean` only have variants of type boolean, and `"flagType":string` flags only have variants of type string, etc, etc... but I don't think we can enforce that ints are ints and not floats (we might be able to do this with [multipleOf](https://json-schema.org/understanding-json-schema/reference/numeric#multiples) but I'd like to see a PoC first).
- Our providers don't care about float/int as it is; they convert instead of indicating a type error.

Very interested in other thoughts on this.

Reply:
The main reason to have both `float` and `int` is to match the flag manifest used in the OpenFeature cli. I think it's important to keep them in sync. However, we may need to support casting from at least int to float.

Reply:
I don't have enough contexts and knowledge to form a mature opinion for OpenFeature spec.... I guess we could explore the extension in flagd and OFREP first and we'll learn more about the feasibility. For now, I think drawing the boundary and aligning on the